### PR TITLE
Rework how we handle custom editors

### PR DESCRIPTION
**Pull request by anthonydresser**
Reworks how we handle using our custom editors (notebook and query editor) to work better in the vscode flow and reduces a lot of the custom edits.

Utilizes serializers to serialize inputs and utilize the editor service overwrite to replace editors when a new file is opened. We still need a custom edit for handling when a user changes the language on an editor.

Not intended for this release.

### CODE REVIEW

**Comment by aaomidi** [comment]
While you're doing this large refactor, can you look into the promise errors?

**Comment by aaomidi** [comment]
I really like this change and it fixes a few personal gripes I've had with the editor. Thanks!

**Comment by anthonydresser** [comment]
Are there specific instances of the promise errors you are thinking of? I don't think there are any specific issues in the code I'm changing here.

**Comment by aaomidi** [comment]
```
WARNING: C:/Users/amomidi/Documents/azuredatastudio/src/sql/workbench/browser/parts/editor/editorStatusModeSelect.ts:45:5 - Promises must be handled appropriately
WARNING: C:/Users/amomidi/Documents/azuredatastudio/src/sql/workbench/browser/parts/editor/editorStatusModeSelect.ts:48:4 - Promises must be handled appropriately

```

Theres a few issues in the other files that have been changed, but they're nothing new. Was curious if it would be worth taking a look at a few of them while you're doing this large refactor.

File: src/sql/workbench/contrib/query/browser/query.contribution.ts [hunk 1]
```diff
@@ -104,7 +116,7 @@ MenuRegistry.appendMenuItem(MenuId.ExplorerWidgetContext, {
 	order: 1
 });
 
-// Query Actions
+// Query Actionsr
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/query.contribution.ts]: nit: extra r added here

File: src/sql/workbench/services/queryEditor/browser/queryEditorService.ts [hunk 2]
```diff
@@ -147,84 +125,10 @@ export class QueryEditorService implements IQueryEditorService {
...
-		const newEditor = await this._editorService.openEditor(newEditorInput, options, group);
-
-		return QueryEditorService._onEditorOpened(newEditor, uri.toString(), undefined, options.pinned);
-	}
-
 	////// Private functions
 
 	private createUntitledSqlFilePath(): Promise<string> {
-		return this.createPrefixedSqlFilePath(untitledFilePrefix);
+		return this.createPrefixedSqlFilePath('SQLQuery');
```
Comment (kburtram) [inline src/sql/workbench/services/queryEditor/browser/queryEditorService.ts]: why are we doing this?

File: src/sql/workbench/browser/parts/editor/editorStatusModeSelect.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+			notificationService.error(localize('languageChangeUnsupported', "Changing editor types on unsaved files is unsupported"));
+			return;
+		}
+		modeSupport.setMode(language);
+		let input: IEditorInput;
+		if (oldInputCreator) {
+			input = oldInputCreator.baseInputCreator(activeEditor);
+		}
+
+		if (newInputCreator) {
```
Comment (kburtram) [inline src/sql/workbench/browser/parts/editor/editorStatusModeSelect.ts]: please add some comments to this method as the logic isn't entirely clear, at least to me.

**Comment by anthonydresser** [comment]
We can revert if any serious issues arise and I'll work through the bugs as I see them.

**Pull request merged**